added more tests for sparse matrices. fail currently. Probably an ind…

lruthotto opened the pull request:
…exing problem?

gaika commented on test/sparse.jl:
```diff
@@ -18,13 +18,28 @@ Ie2 = full(Aid)
+@test all(Ah .== SparseMatrixCSC(AH))
+(V,R,C,t) = sparse_get_info(AH)
+@test norm(Array(V) - Ah.nzval,Inf) < 1e-7
+@test all(R.==Ah.rowval)
```
AFArray internal representation for indexes is 0 based

lruthotto commented:
I fixed the test file now and all tests pass on my machine. I ran into some problems with the conversion to SparseMatrixCSC. I think the code that is in util.jl is not correct, but was unable to fix it so that it works for all cases. 

I replaced line 149 in util.jl by
```
SparseMatrixCSC(sz[2], sz[1], Array(rowval+1), Array(colptr+1), Array(nzval))'
```
since I assume that the rowval and colptr returned by sparse_get_info are in CSR format. Am I wrong?

lruthotto commented:
Here's something to try

```
@test all(speye(10) .== SparseMatrixCSC(AFArray(speye(10))))
```
this should work, but 

```
Ah = sprandn(2*n,n,.1)
AH = sparse(AFArray(full(Ah)))
@test all(Ah .== SparseMatrixCSC(AH))
```
fails. Note that the construction of `AH` is necessary to make other things work. For example, `AH= AFArray(Ah)` causes `sparse_get_nnz(AH)` to return something else than `nnz(Ah)`. I'm a little confused by this, so I'd be happy to get some thoughts.

gaika commented:
I think it ends up being transposed, rows / columns are mixed up somewhere.

gaika commented:
AF_STORAGE_CSC is not supported:

In file src/api/c/sparse.cpp:190
Storage type is out of range/unsupported

lruthotto commented:
Thanks for the responses. 

I think it is very sensible to have CSR matrices on the GPU since matvecs are more efficient and easier to parallelize, so I don't think a change to ArrayFire would make sense. 

This means though that we need to be extra careful in this package when converting `SparseMatrixCSC` to `AFArray` and vice versa and with functions like `full` and `sparse` called on `AFArray`. I've found a few more things that are not consistent. Try this:

```
A = sprandn(10,10,.1);
display(full(A))
AF = AFArray(A)
display(full(AF))  # this looks incorrect
display(full(SparseMatrixCSC(AF))) # looks correct
```

lruthotto commented:
For `full` and `sparse` I'd need your help. They seem to get executed using the wrapper. Could there be a bug?

gaika commented:
If I transpose CSC before converting to CSR and make sure indexes are Int32 everything seem to work with 0a5a8218a34d7a8dc48d79598d6625eab8ad412a
